Добавлено поддержка Autofac

bastonta opened the pull request:
@mitro, @devmankz переписал проверьте

bastonta committed:
Переписано репозитории с использование паттерна Unit Of Work

bastonta committed:
Убраны дубликаты кода в репозиториях

bastonta committed:
Добавлено поддержка Autofac

duskembayev commented on Depmon.Server.Collector.App/config.xml:
```diff
@@ -0,0 +1,7 @@
+  <mailboxes>
+    <add name="hm@ecm.local" server="sts.ecm.local" port="110" ssl="false" username="hm@ecm.local" password="1" timeout="1000" capacity="10" delay="00:00:10" />
+  </mailboxes>
+</monitoring>
```
Конфиг надо вернуть в app.config.

bastonta replied:
исправлено

duskembayev commented on Depmon.Server.Collector.Impl/DataLoad.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace Depmon.Server.Collector.Impl
+{
+    public class DataLoad : IDataLoad
```
Предлагаю переименовать класс. Название ни о чем не говорит.
MailReciever какой нить или типа.

mitro replied:
Да, и класс всегда должен называться именем существительным. А метод - глаголом.

bastonta replied:
исправлено

duskembayev commented on Depmon.Server.Collector.Impl/DataSave.cs:
```diff
@@ -0,0 +1,79 @@
+
+                using (var scope = continer.BeginLifetimeScope())
+                {
+                    using (var unitOfWork = scope.Resolve<IUnitOfWork>())
```
Раз используем ioc - надо использовать все его прелести.
Прокидывать все зависимости через конструктор.

mitro replied:
Да, просто репозитории инжектировать

bastonta replied:
Исправлено

duskembayev commented on Depmon.Server.Collector.Impl/DataSave.cs:
```diff
@@ -0,0 +1,79 @@
+            {
+                IContainer continer = new AutofacContainer().GetContainer();
+
+                using (var scope = continer.BeginLifetimeScope())
```
Scope создается один раз за итерацию. В каждом потоке свой. Один и только один.

bastonta replied:
исправлено

duskembayev commented on Depmon.Server.Collector.Impl/DataSave.cs:
```diff
@@ -0,0 +1,79 @@
+
+namespace Depmon.Server.Collector.Impl
+{
+    public class DataSave : IDataSave
```
Тоже надо переменовать

duskembayev replied:
Еще разделить логику парсинга и сохранения на разные классы. Вынести вызовы этих процессов на уровень итераторов. Там в тему будет создание scope.

bastonta replied:
исправлено

duskembayev commented on Depmon.Server.Collector.Impl/LevelConverter.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Depmon.Server.Collector.Impl
+{
+    class LevelConverter : ILevelConverter
```
Предлагаю вернуть как  internal class в FactMap, т.к. является неотьемлемой ее частью.

bastonta committed:
Удалил класс ConnectionFactory

bastonta committed:
Переименовал класс DataLoad в MailReceiver

mitro commented on Depmon.Server.Collector.Impl/Configuration/ConfigReader.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    public class ConfigReader : IConfigReader
+    {
+        private const string CONFIG_FILE = "config.xml";
```
Используй `PascalCase` для именования констант в C# http://stackoverflow.com/a/242549/1262176

bastonta replied:
исправлено

bastonta committed:
Переименовал константу используя PascalCase

duskembayev commented on Depmon.Server.Collector.Impl/AutofacContainer.cs:
```diff
@@ -0,0 +1,30 @@
+        {
+            var builder = new ContainerBuilder();
+
+            builder.RegisterType<UnitOfWork>().As<IUnitOfWork>();
```
Должен создавать один раз за один Scope. А сейчас будет создаваться каждый раз, при обращении.
Здесь же должна быть инициализация.

mitro commented on Depmon.Server.Collector.Impl/Configuration/ConfigReader.cs:
```diff
@@ -0,0 +1,42 @@
+            }
+            
+            return result;
+        }
```
Если здесь использовать подход с созданием классов, представляющий секцию, код не будет проще? Определить классы, десериализовать в них конфиг, а тут просто скопировать полученные значения в `result`. Чтобы руками не парсить

duskembayev commented on Depmon.Server.Collector.App/Program.cs:
```diff
@@ -0,0 +1,39 @@
+            Console.WriteLine("\nPress 'X' for break...\n\n");
+
+            IContainer container = new AutofacContainer().GetContainer();
+            using (var scope = container.BeginLifetimeScope())
```
Нужно хорошо продумать модель создания Scope.
Один Scope - один uow. 
Если создаем scope в рамках одного потока - то соединение к БД будет постоянно открытым, а так нельзя делать. Значит, надо scope открывать на каждом проходе каждого потока. Это на самом деле нормально.

mitro commented on Depmon.Server.Collector.Impl/DataLoad.cs:
```diff
@@ -0,0 +1,59 @@
+                client.Connect(mailbox.Server, mailbox.Port, mailbox.Ssl, mailbox.Timeout, mailbox.Timeout, CertificateValidator);
+                client.Authenticate(mailbox.Username, mailbox.Password);
+
+                var lCount = client.GetMessageCount();
```
Не используй префиксы, описывающие тип переменной. Это называется _hungarian notation_ и в C# не используется

duskembayev replied:
Вообще тут имелось ввиду letterCount, что тоже не очень удачно. :)

mitro commented on Depmon.Server.Collector.Impl/Engine.cs:
```diff
@@ -0,0 +1,76 @@
+        {
+            Console.WriteLine("[{0}] monitoring started", mailbox.Name);
+
+            IContainer continer = new AutofacContainer().GetContainer();
```
В правильно написанном коде зависимость на Autofac должна быть только в корне приложения. В остальные классы зависимости должны прокидываться через конструктор. Это должно работать в большинстве случаев

mitro commented on Depmon.Server.Collector/Configuration/Mailbox.cs:
```diff
@@ -1,36 +1,25 @@
+
+namespace Depmon.Server.Collector.Configuration
+{
+    public class Mailbox
```
Предлагаю для всех классов, представляющих данные конфигурации, использовать паттерн "Value object". Это позволит обезопасить себя от того, что свойства объекта будут изменены после его создания. Почитай про паттерн и реализуй

mitro commented on Depmon.Server.Collector/IDataLoad.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Depmon.Server.Collector
+{
+    public interface IDataLoad
```
Сделай имя существительное

bastonta replied:
переименовал

mitro commented on Depmon.Server.Database/FactRepository.cs:
```diff
@@ -1,22 +1,33 @@
+        public FactRepository(IDbConnection connection) : base(connection)
+        { }
+
+        public override void Save(Fact entity)
```
В классах, реализующих интерфейс, ты можешь переименовывать параметры методов. Переименуй `entity` в `fact`, чтобы код методов проще читался. Ну и в остальных местах по аналогии

bastonta replied:
исправил

mitro commented on Depmon.Server.Database/IRepository.cs:
```diff
@@ -0,0 +1,18 @@
+
+        T GetById(int id);
+
+        void InsertMany(params T[] entity);
```
Используй множественное число `entities`

bastonta replied:
исправил

mitro commented on Depmon.Server.Collector/Configuration/MonitoringSection.cs:
```diff
@@ -1,13 +1,11 @@
+
+namespace Depmon.Server.Collector.Configuration
+{
+    public class MonitoringSection
```
Может переименовать в `Settings`? А то ты абстрагировал место, откуда вытаскиваются настройки, а в названии класса намек на `app.config` остался.

duskembayev commented on Depmon.Server.Database/IUnitOfWork.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IUnitOfWork : IDisposable
+    {
+        IRepository<T> GetRepository<T>();
```
IUnitOfWork - это не фабрика репозиториев. Это владелец соединение к БД (если проще). Почитай про паттерны: UnitOfWork и Repository.

bastonta replied:
исправлено

duskembayev commented on Depmon.Server.Database/UnitOfWork.cs:
```diff
@@ -0,0 +1,51 @@
+                object repository;
+
+                //TODO: придумать другой вариант
+                switch (type)
```
Используй Autofac для создания и управления репозиториями.

bastonta replied:
исправлено

bastonta committed:
Исправлено наименование

bastonta committed:
Вернул настройки в app.config

bastonta committed:
Вернул класс LevelConverter в FactMap

bastonta committed:
Разделено париснг csv и сохранение фактов

bastonta committed:
Исправлено название локальной переменной

bastonta committed:
Правельно реализовано UoW

bastonta committed:
Autofac зависимости передаются через конструктор

bastonta committed:
UoW создается в каждом потоке свой

bastonta committed:
Убрал статик

bastonta committed:
Connection string передается через параметр

bastonta committed:
Scope создается в каждой итерации

bastonta committed:
Метод Commit вызывается у объекта, который возвращается при вызове метода BeginTransaction

bastonta committed:
Заменен разделитель в csv файлах

bastonta committed:
Удалил закомментированный код

bastonta committed:
Обернул в using

bastonta committed:
Scope контролирует тот кто его запросил

bastonta committed:
Раделитель csv исправил на |

bastonta committed:
Убрал ссылку ссылку для transaction у UoW

bastonta committed:
Убрал ссылки на transaction

bastonta committed:
Stream закрываю

bastonta committed:
Не нужный try catch

bastonta committed:
Уничтажаю data

bastonta committed:
TableName сделал абстракным свойством

bastonta committed:
Добавил публичное свойство Session

bastonta committed:
Инжектировал IUnitOfWork в Repository

bastonta committed:
Autofac конфиг исправил

duskembayev commented on Depmon.Server.Collector.App/Program.cs:
```diff
@@ -1,33 +1,42 @@
+            {
+                cr = scope.Resolve<IConfigReader>();
+                engine = scope.Resolve<IEngine>();
+            }
```
Ты убил `scope`, а порожденные от него объекты продолжают жить. Ты либо время жизни `scope` увеличь, либо порождай эти объекты напрямую от контейнера. Мне кажется правильным - второй вариант.

duskembayev commented on Depmon.Server.Collector.Impl/CsvParse.cs:
```diff
@@ -0,0 +1,39 @@
+            }
+            catch (Exception e)
+            {
+                throw e;
```
Какой смысл к `try...catch`?

bastonta committed:
Engine и ConfigReader создаются от контейнера

duskembayev commented on Depmon.Server.Collector.Impl/Engine.cs:
```diff
@@ -0,0 +1,97 @@
+                {
+                    IMailReciever mailReciever = scope.Resolve<IMailReciever>();
+                    IFactsSave factsSave = scope.Resolve<IFactsSave>();
+                    ICsvParse csvParse = scope.Resolve<ICsvParse>();
```
Переименуй в `ICsvReader`. Не забудь переменную тоже переименовать.

duskembayev commented on Depmon.Server.Collector.Impl/Engine.cs:
```diff
@@ -0,0 +1,97 @@
+                using (var scope = _objectFactory.CreateScope())
+                {
+                    IMailReciever mailReciever = scope.Resolve<IMailReciever>();
+                    IFactsSave factsSave = scope.Resolve<IFactsSave>();
```
Переименуй ReportRegistry

bastonta committed:
Переименовал класс в CsvReader

bastonta committed:
Переименовал класс в ReportRegistry

duskembayev commented on Depmon.Server.Collector.Impl/FactsSave.cs:
```diff
@@ -0,0 +1,40 @@
+            {
+                var report = new Report {CreatedAt = DateTime.Now};
+                _reportRepository.Save(report);
+                var reportId = _reportRepository.GetAll().FirstOrDefault(s => s.CreatedAt == report.CreatedAt).Id;
```
У тебя метод `Save()` должен был заполнить поле `report.Id`. Тогда не пришлось бы делать еще один запрос.
Представь что было бы на этой строчке, если в таблице `Reports` будет **1000000** записей?

duskembayev commented on Depmon.Server.Collector.Impl/FactsSave.cs:
```diff
@@ -0,0 +1,40 @@
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"Error: {e.Message}");
```
Не просто **Error**, а нормально надо писать - при каком действии. Например:
`Report saving failed: {e.message}`.

duskembayev commented on Depmon.Server.Collector/Configuration/IConfigReader.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Depmon.Server.Collector.Configuration
+{
+    public interface IConfigReader
```
Переименуй `ISettings`

duskembayev commented on Depmon.Server.Collector/IObjectFactory.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Depmon.Server.Collector
+{
+    /// <summary>
```
Убери описание, которое не соответствует действительности.

bastonta committed:
Убрал лишний try catch

bastonta committed:
Переименовал класс в Settings

bastonta committed:
Убрал комментарии

duskembayev commented on Depmon.Server.Database/UnitOfWork.cs:
```diff
@@ -0,0 +1,40 @@
+            Dispose(true);
+        }
+
+        protected void Dispose(bool disposing)
```
Зачем тебе это метод? Убери этот метод и логику перенеси в `Dispose`.

duskembayev commented on Depmon.Server.Database/Repository.cs:
```diff
@@ -0,0 +1,46 @@
+        protected abstract string TableName { get; }
+        protected IUnitOfWork _unitOfWork;
+
+        public Repository(IUnitOfWork unitOfWork)
```
Конструктор у абстрактного класса не может быть `public`.

bastonta committed:
При сохранений возвращается id

bastonta committed:
Убрал лишний метод

bastonta committed:
Убрал public у конструктора абстрактного класса

mitro commented:
Добавь в `Report` коллекцию `Facts`

mitro commented on Depmon.Server.Collector.Impl/Engine.cs:
```diff
@@ -0,0 +1,96 @@
+                    {
+                        data = mailReciever.Load(mailbox);
+
+                        foreach (var stream in data)
```
Обернуть в транзакцию, полученную из `IUnitOfWork`

devmankz commented:
А зачем коллекция `Facts` в `Report`?
Даже если надо будет - сделаем, когда появиться необходимость.

ablai96 commented:
ее не надо добавлять сказали

bastonta committed:
Исправил заметки @mitro

- Убрал название конкретной таблицы
- Переименовал класс в SettingsReader и Settings
- AutofacContainer перенес в сборку Depmon.Server.Collector.App
- В ICsvReader метод Parse переименовал в Read
- Добавил транзакцию при сохранении Report и Fact
